x-pack/plugins/security/server/config_deprecations.ts

```diff
@@ -136,22 +136,25 @@ export const securityConfigDeprecationProvider: ConfigDeprecationProvider = ({
     }
   },
   (settings, fromPath, addDeprecation) => {
-    if (settings?.xpack?.security?.enabled === false) {
+    if ('enabled' in (settings?.xpack?.security || {})) {
       addDeprecation({
         title: i18n.translate('xpack.security.deprecations.enabledTitle', {
-          defaultMessage: 'Disabling the security plugin "xpack.security.enabled" is deprecated',
+          defaultMessage: 'Setting "xpack.security.enabled" is deprecated',
         }),
         message: i18n.translate('xpack.security.deprecations.enabledMessage', {
           defaultMessage:
-            'Disabling the security plugin "xpack.security.enabled" will only be supported by disable security in Elasticsearch.',
+            'Enabling or disabling the security plugin from Kibana using "xpack.security.enabled" is deprecated. This should instead be controlled via Elasticsearch.',
         }),
+        documentationUrl:
+          'https://www.elastic.co/guide/en/kibana/current/security-settings-kb.html#general-security-settings',
```

Comment:
Once v8.0.0 is released `current` is going to refer to this version, in which we'll probably remove the section of the documentation we're linking to here. Shouldn't we instead update this link to be "fixed" to the `7.x` branch somehow? For example:

```suggestion
          'https://www.elastic.co/guide/en/kibana/7.x/security-settings-kb.html#general-security-settings',
```

The downside of this is that it will show this warning in the top of the page:

![image](https://user-images.githubusercontent.com/10602/132642583-a232e871-a35c-4925-9f99-02a6ed2f5096.png)

Alternatively we could set it to `7.16` instead, but that will be 404 until that version is released 🤷

Reply:
Hm, normally I'd say you should use the [DocLinksService](https://github.com/elastic/kibana/blob/e16736416a14e48e1c6578c9377982dbe4124cf0/src/core/public/doc_links/doc_links_service.ts), but there is no server-side counterpart for that plugin.

It looks like this just uses the `branch` from the `packageInfo` to generate these URLs. You can get this in the security plugin like this:
```diff
diff --git a/x-pack/plugins/security/server/plugin.ts b/x-pack/plugins/security/server/plugin.ts
index 1873ca42324..85a21b42074 100644
--- a/x-pack/plugins/security/server/plugin.ts
+++ b/x-pack/plugins/security/server/plugin.ts
@@ -188,6 +188,7 @@ export class SecurityPlugin
 
   constructor(private readonly initializerContext: PluginInitializerContext) {
     this.logger = this.initializerContext.logger.get();
+    const branch = this.initializerContext.env.packageInfo.branch;
   }
 
   public setup(
```

then you'll need to pass that variable down to the config deprecations to use it there.

Reply:
The URLs need to be pinned to the branch. Hardcoding links to master or 7.x is like linking to quicksand. 

I'm concerned that we have links that are hardcoded in place. Is it possible to move this (and similar links) to a resource file so that we can test them as part of the doc build? Otherwise, it's entirely possible for doc changes to cause links like this that appear in Kibana to 404.

Reply:
I don't think it's worth linking to this topic, TBH. It doesn't tell them anything about what they need to do. Sending them to the top of the Security docs would be more helpful. https://www.elastic.co/guide/en/elasticsearch/reference/7.x/secure-cluster.html

@lockewritesdocs is actively working on the story around Security being on by default in 8.0. This might be an area where we need to provide explicit info about handling the changes between 7.16 and 8.0. (Which would then be an ideal target for this link.)

Reply:
My intent with this link was to send users to a doc explaining how to configure Elasticsearch to do what the deprecated config option in Kibana is doing. I know that technically I link to the Kibana docs and that you have to click a link there to go to the corresponding Elasticsearch docs, but I thought it would then be telling the whole story that way (this is the page they would end up on if they followed the link in the Kibana docs: https://www.elastic.co/guide/en/elasticsearch/reference/7.14/security-settings.html#general-security-settings)

Reply:
However, you could argue that the link is irrelevant as we explain in the deprecation message that they should set the equivalent config option in Elasticsearch as well. And if they know how to do that they shouldn't need more information.

Reply:
> I don't think it's worth linking to this topic, TBH. It doesn't tell them anything about what they need to do. Sending them to the top of the Security docs would be more helpful. https://www.elastic.co/guide/en/elasticsearch/reference/7.x/secure-cluster.html
> 
> @lockewritesdocs is actively working on the story around Security being on by default in 8.0. This might be an area where we need to provide explicit info about handling the changes between 7.16 and 8.0. (Which would then be an ideal target for this link.)

I agree with @debadair. Linking to the top-level page is the better, more useful option. I updated [this page](https://www.elastic.co/guide/en/elasticsearch/reference/master/secure-cluster.html) for alpha1, and will be making more changes for alpha2 next week. On this page, we'll be informing users that security is enabled by default in 8.0. If they're coming from 7.x:
* Security is explicitly configured: users can continue using that setup.
* Security is explicitly **disabled** -- meaning they set `xpack.security.enabled` to `false`: users can continue running without security at their own peril. That's why I indicated in [another comment](https://github.com/elastic/kibana/pull/111676/files/ed7a44064f36f8718a8a96294b3542f571ce1486#r706422263) that we should inform users to explicitly indicate a value for this setting. 
* Security is not configured -- meaning we didn't find a value for `xpack.security.enabled`: we're going to enable security for them. From the [design doc](https://docs.google.com/document/d/1KUzflsuq4Ixz0EgRNBORRgqEEbpf8-nFFn7gZDHw9_8/edit#heading=h.v8d59avlewid):
>we believe that forcing that decision explicitly upon them while they upgrade is beneficial as it will nudge a subset of these users to enable and configure security before they upgrade. For the rest of the users that will decide to disable security in order to upgrade, this becomes an explicit decision and there is a better chance that they will attempt to amend this by enabling security at some point, now that they are made aware.

Reply:
@lockewritesdocs we were thinking we should link to the 7.16 docs since this will be in the 7.16 upgrade assistant, are you suggesting we should link to the 8.0 docs instead? 

Alternatively, do you think it makes sense to update that page for 7.16 as well?

Reply:
@jportner -- apologies, I crossed my wires on this one. Been working on 8.0 stuff. Yes, for 7.16 we should link to the 7.16 version of that page, which we will update to indicate that security is **not** enabled by default, and that users should explicitly configure it. We're updating all of the product Download pages for 7.15 with this same message, and a link to the docs for how to configure security. Security wasn't ever on by default (even before 7.15), but we basically want to say, "Security isn't enabled. We suggest that you enable it, because that keeps your data and anything connected to Elasticsearch safer. Here's a link to go enable security, unless you **really** don't want to." I'm going to get a PR in place for that change now and mark it for 7.15, and also apply to 7.x.

Reply:
I suggest that I update this URL to point to what we want in "current" and then make an issue to expose the current branch to `ConfigDeprecationProvider` functions in general as this isn't something that's only needed for this file. Then that will not be a blocker for this PR - and since we already have several URLs in this file that link to "current", we're not really making it any worse by adding this one either.

Reply:
Ok, according to my comment above, I've now updated the URL to be:

> https://www.elastic.co/guide/en/elasticsearch/reference/current/secure-cluster.html

Reply:
We spoke offline but I wanted to mention here that

> I suggest that I update this URL to point to what we want in "current" and then make an issue to expose the current branch to `ConfigDeprecationProvider` functions in general as this isn't something that's only needed for this file. Then that will not be a blocker for this PR - and since we already have several URLs in this file that link to "current", we're not really making it any worse by adding this one either.

:+1: :+1: :+1:

Reply:
I've created an issue for it now: https://github.com/elastic/kibana/issues/112221

I wasn't sure which team to assign it to, so I've left that blank for now.